# Web Intelligence
## Python crash course
## Pandas Library (and Numpy)

#### Prof. Claudio Lucchese

## NumPy 

http://www.numpy.org/

**NumPy** stands for **Numerical Python**.

NumPy is a fundamental package for efficient scientific and numerical computing.

It provides:
- efficient methods for managing arrays and matrices, and operations on them
- several mathematical functions (variance, standard deviation, cumulative sum, ...)
- other: fitting a polynomial, finding the minimum of a function, Fast Fourier Transform, etc.


## Matrix representation and operations

In [3]:
# transform a nested list into a 2D matrix

import numpy as np
a = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print ("Matrix a:")
print (a)

Matrix a:
[[1. 2. 3.]
 [4. 5. 6.]]


In [5]:
# Element-wise operations OPERAZIONI ELEMENTO PER ELEMENTO
b = a + 2
print ("Matrix b=a+2:")
print (b)
print()

c = a * 7
print ("Matrix c=a*7:")
print (c)
print()

d = b - a
print ("Matrix d=b-a:")
print (d)
print()

e = c / a
print ("Matrix e=c/a:")
print (e)

Matrix b=a+2:
[[3. 4. 5.]
 [6. 7. 8.]]

Matrix c=a*7:
[[ 7. 14. 21.]
 [28. 35. 42.]]

Matrix d=b-a:
[[2. 2. 2.]
 [2. 2. 2.]]

Matrix e=c/a:
[[7. 7. 7.]
 [7. 7. 7.]]


In [6]:
# Matrix transpose and multiplication MOLTIPLICO PRIMA MATRICE X TRASPOSTA DELLA SECONDA
z = np.matmul(a, b.T)
print (z)
print (b.T)

[[ 26.  44.]
 [ 62. 107.]]
[[3. 6.]
 [4. 7.]
 [5. 8.]]


## Mathematical functions

See https://docs.scipy.org/doc/numpy/reference/ufuncs.html for more information.

Let's test `sqrt()` (square root) and `maximum()` (element-wise maximum of two arrays).

In [7]:
a = np.array( [ [1.,2.,3.],[4.,5.,6.] ] )
b = np.array( [ [3.,3.,7.],[1.,1.,2.] ] )

print ("exp function")
print ( np.exp(a))
print ()

print ("element-wise maximum") # FATTA ELEMENTO X ELEMENTO
print ( np.maximum(a,b) )

exp function
[[  2.71828183   7.3890561   20.08553692]
 [ 54.59815003 148.4131591  403.42879349]]

element-wise maximum
[[3. 3. 7.]
 [4. 5. 6.]]


## Aggregation and Statistical methods

Typical `sum`, `mean`, `var`, are available ...

In case of matrices, it is possible to specify the **direction** of the operation.
 - see https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html#numpy.mean

In [8]:
m = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print (m)
print ()

print ("No direction/axis")
print ( np.mean(m) ) #MEDIA DI TUTTI I NUMERI
print ()

print ("mean over axis 0 (across rows)")
print ( np.mean(m, axis=0) ) #MEDIA DI NUMERI PER OGNI COLONNA ATTRAVERSANDO LE RIGHE
print ()

print ("mean over axis 1 (across cols)") #MEDIA PER I NUMRI DI OGNI RIGA ATTRAVERSANDO TUTTE LE COLONNE
print ( np.mean(m, axis=1) )
print ()


[[1. 2. 3.]
 [4. 5. 6.]]

No direction/axis
3.5

mean over axis 0 (across rows)
[2.5 3.5 4.5]

mean over axis 1 (across cols)
[2. 5.]



## Indexing and Views

It is possible to access elements of matrices/array in a similar way to Python lists.

Slices of a matrix are implemented as **views**, **not copies**, on the original data, sharing the same memory.

In [9]:
m = np.array( [ [1.,2.,3.],
                [4.,5.,6.] ] )

print (m)
print ()

print ("view/indexing")
view = m[:,1]  # allo rows, column with index 1
print ( view )
print ()

print ("modified view and original data")
view[:] = 33
print ( view )
print ()
print ( m )    # also m is modified

[[1. 2. 3.]
 [4. 5. 6.]]

view/indexing
[2. 5.]

modified view and original data
[33. 33.]

[[ 1. 33.  3.]
 [ 4. 33.  6.]]


## Sorting and Fancy Indexing

Similarly to python, the sort method can be used to sort an array or to get a sorted copy.

There is no `key` parameter.

See https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.sort.html.

In [10]:
data = np.array([1,-2,3,-4,5])

s = np.sort(data)
print ( "sorted copy    ", s )
print ( "original data  ", data)
print()

data.sort()
print ( "after data.sort()" )
print ( "original data  ", data)

sorted copy     [-4 -2  1  3  5]
original data   [ 1 -2  3 -4  5]

after data.sort()
original data   [-4 -2  1  3  5]


Also sorting may have a direction.

In [11]:
m = np.array( [ [3., 5.,2.],[6., 1.,5.] ] )

print (m)

print ("sort over axis 0 (across rows)")
print ( np.sort(m, axis=0) )
print()

print ("sort over axis 1 (across cols)")
print ( np.sort(m, axis=1) )
print()

print ("flatten and then sort")
print ( np.sort(m, axis=None) )

[[3. 5. 2.]
 [6. 1. 5.]]
sort over axis 0 (across rows)
[[3. 1. 2.]
 [6. 5. 5.]]

sort over axis 1 (across cols)
[[2. 3. 5.]
 [1. 5. 6.]]

flatten and then sort
[1. 2. 3. 5. 5. 6.]


A useful method is `argsort`, which returns the positions of the elements in sorted order, withouth modifying the original array.

The output of `argsort`, can be used in conjuction with fancy indexing.

Example: sort by income and print the corresponding name.

In [12]:
names = np.array([  'Mark', 'Joe', 'Will', 'Bob', 'Jane', 'Carol', 'Donald'])
salaries = np.array([2000,   1200,  3000,  2100,   1580,   1700,    900])

sorted_pos = np.argsort(salaries) #DICE IN CHE POSIZIONE SONO I DATI ORDINATI
print ("sorted positions", sorted_pos)

print ( "ages",  names[sorted_pos] ) # This is called fancy indexing!

sorted positions [6 1 4 5 0 3 2]
ages ['Donald' 'Joe' 'Jane' 'Carol' 'Mark' 'Bob' 'Will']


## Optimization

Find the minimum of a function given it first derivative.

See https://docs.scipy.org/doc/scipy/reference/optimize.html

In [9]:
from scipy import optimize
def f(x):
    return (x**3 - 1)  # only one real root at x = 1

def fprime(x):
    return 3*x**2

sol = optimize.root_scalar( f,                # function
                            x0=0.2,           # initial guess
                            fprime=fprime,    # first derivative
                            method='newton')  # optimization method

print ("The root of the function is:", sol.root)

The root of the function is: 1.0


# More Numpy ?

NumPy deserves your interest, especially for implementing numerical algorithms, matrix-based operations, and to exploit its great algorithms (see also scipy).

From a data perspective, it provides a low level access. We will see the Pandas library, which is more data-oriented and it shares several commonalities with NumPy.

## Pandas: Python Data Analysis Library

See: https://pandas.pydata.org/

Pandas is an open source providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.

## Let's read our dataset with one line of code!

We can read an excel file!

See: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html

Check also the `sheet_name` parameter.

In [13]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'

df = pd.read_excel(dataset_file) 
df.head() # see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html

ATP  Location              Tournament       Date  Series    Court Surface  \
0    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
1    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
2    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
3    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
4    1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   

       Round  Best of         Winner  ... Lsets    Comment  B365W  B365L  \
0  1st Round        3    Dimitrov G.  ...   0.0  Completed   1.36   3.00   
1  1st Round        3      Raonic M.  ...   0.0  Completed   1.18   4.50   
2  1st Round        3  Kecmanovic M.  ...   0.0  Completed   1.57   2.25   
3  1st Round        3     Millman J.  ...   1.0  Completed   1.40   2.75   
4  1st Round        3    Uchiyama Y.  ...   0.0  Completed   2.62   1.44   

    PSW   PSL  MaxW  MaxL  AvgW  AvgL  
0  1.36  3.37  1.42  3.60  1.35  3.18  
1  1.23  4.68  1.27  4.84  1.22  4.26  
2  1.67  2.32  1.71  2.40  1.63  2.28  
3  1.41  3.13  1.45  3.20  1.40  2.95  
4  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

In [11]:
print (type(df))
# see https://pandas.pydata.org/pandas-docs/stable/reference/frame.html

<class 'pandas.core.frame.DataFrame'>


## Dataframe

- **DataFrame** is a 2-dimensional table of data and contains an ordered collection of columns, each of which can be a different value type (numeric, string, boolean, etc.).

See: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html

**Features of DataFrame:**
- Mutable
- Potentially columns are of different types
- Labeled axes (rows and columns)
- Can Perform Arithmetic operations on rows and columns

A DataFrame can be seen as a dictionary of columns (indeed, pandas Series), all sharing the same index. 

We can ask the number of rows and columns.
We can access index and columns labels with attributes `index`and `column`.

In [14]:
print("df.shape:", df.shape) #no righe, no colonne
print("df.index:", df.index) #nomi righe
print("df.columns:", df.columns) #nomi colonne

df.shape: (2234, 36)
df.index: RangeIndex(start=0, stop=2234, step=1)
df.columns: Index(['ATP', 'Location', 'Tournament', 'Date', 'Series', 'Court', 'Surface',
       'Round', 'Best of', 'Winner', 'Loser', 'WRank', 'LRank', 'WPts', 'LPts',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'Comment', 'B365W', 'B365L', 'PSW', 'PSL', 'MaxW', 'MaxL',
       'AvgW', 'AvgL'],
      dtype='object')


## Question: What is the number of matches?


In [15]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 

print ("Dataframe shape is:", df.shape)
num_matches, num_columns = df.shape
print ("The number of matches is:", num_matches)

Dataframe shape is: (2234, 36)
The number of matches is: 2234


### Names

Both index and columns have an attribute `name` to specify their names.


In [16]:
df.index.name   = "Match ID"
df.columns.name = "Attributes"

df.head()

Attributes  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                        
0             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Attributes Surface      Round  Best of         Winner  ... Lsets    Comment  \
Match ID                                               ...                    
0             Hard  1st Round        3    Dimitrov G.  ...   0.0  Completed   
1             Hard  1st Round        3      Raonic M.  ...   0.0  Completed   
2             Hard  1st Round        3  Kecmanovic M.  ...   0.0  Completed   
3             Hard  1st Round        3     Millman J.  ...   1.0  Completed   
4             Hard  1st Round        3    Uchiyama Y.  ...   0.0  Completed   

Attributes  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Match ID                                                      
0            1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
1            1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
2            1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
3            1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
4            2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

## Statistical summary

We can get a statistical summary for numerical columns.

In [15]:
df.describe()

Attributes          ATP      Best of        WRank        LRank          WPts  \
count       2234.000000  2234.000000  2230.000000  2221.000000   2231.000000   
mean          28.635184     3.453894    57.948430    79.944169   1723.479606   
std           15.505220     0.837904    54.876823    76.409609   2022.360691   
min            1.000000     3.000000     1.000000     1.000000     17.000000   
25%           17.000000     3.000000    20.000000    36.000000    707.000000   
50%           29.000000     3.000000    48.000000    64.000000    983.000000   
75%           41.000000     3.000000    78.000000    98.000000   1740.000000   
max           54.000000     5.000000   455.000000  1415.000000  12415.000000   

Attributes          LPts           W1           L1           W2           L2  \
count        2221.000000  2215.000000  2215.000000  2205.000000  2205.000000   
mean         1104.742008     5.823928     4.190971     5.822676     4.043537   
std          1145.908936     1.199468     1.828982     1.229490     1.840488   
min             3.000000     0.000000     0.000000     0.000000     0.000000   
25%           577.000000     6.000000     3.000000     6.000000     3.000000   
50%           830.000000     6.000000     4.000000     6.000000     4.000000   
75%          1205.000000     6.000000     6.000000     6.000000     6.000000   
max         12355.000000     7.000000     7.000000     7.000000     7.000000   

Attributes  ...        Wsets        Lsets        B365W        B365L  \
count       ...  2215.000000  2215.000000  2225.000000  2225.000000   
mean        ...     2.200451     0.445147     1.850387     3.222966   
std         ...     0.455321     0.577775     0.917769     3.228215   
min         ...     0.000000     0.000000     1.002000     1.070000   
25%         ...     2.000000     0.000000     1.300000     1.660000   
50%         ...     2.000000     0.000000     1.570000     2.300000   
75%         ...     2.000000     1.000000     2.100000     3.500000   
max         ...     3.000000     2.000000     9.000000    41.000000   

Attributes          PSW          PSL         MaxW         MaxL         AvgW  \
count       2225.000000  2225.000000  2234.000000  2234.000000  2234.000000   
mean           1.937683     3.526445     2.002936     3.743693     1.871629   
std            1.051800     3.723625     1.125870     4.409477     0.933585   
min            1.005000     1.070000     1.010000     1.080000     1.010000   
25%            1.330000     1.760000     1.360000     1.820000     1.310000   
50%            1.640000     2.440000     1.670000     2.510000     1.600000   
75%            2.190000     3.650000     2.270000     3.800000     2.120000   
max           11.730000    37.800000    12.220000    67.000000     9.640000   

Attributes         AvgL  
count       2234.000000  
mean           3.186124  
std            2.769056  
min            1.060000  
25%            1.720000  
50%            2.340000  
75%            3.477500  
max           28.490000  

[8 rows x 26 columns]

## Data Types

A specific data type is used to store and manage the information in the dataframe. This is important to understand which operations can be performed on the different columns.

Note that also the non-null values are reported. It is not uncommon to have missing values in our dataset.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2234 entries, 0 to 2233
Data columns (total 36 columns):
ATP           2234 non-null int64
Location      2234 non-null object
Tournament    2234 non-null object
Date          2234 non-null datetime64[ns]
Series        2234 non-null object
Court         2234 non-null object
Surface       2234 non-null object
Round         2234 non-null object
Best of       2234 non-null int64
Winner        2234 non-null object
Loser         2234 non-null object
WRank         2230 non-null float64
LRank         2221 non-null float64
WPts          2231 non-null float64
LPts          2221 non-null float64
W1            2215 non-null float64
L1            2215 non-null float64
W2            2205 non-null float64
L2            2205 non-null float64
W3            1119 non-null float64
L3            1119 non-null float64
W4            265 non-null float64
L4            265 non-null float64
W5            96 non-null float64
L5            96 non-null float64
Wset

## Column selection, addition, and deletion

See: https://pandas.pydata.org/pandas-docs/stable/indexing.html.

Pandas provides several (sometimes confusing) ways to access the columns of a data frame.

#### Example (Selection)
We can select a column by specifying its name. 

This operation gives us Pandas Series.

Note that the index is preserved.

In [17]:
a = df["Winner"] #COSI' STO CHIEDENDO UNA COLONNA

print("a:\n", a)
print("Type: ", type(a))
print()

# Equivalent to

a = df.Winner #ALTRO MODO EQUIVALENTE PER ACCEDERE A COLONNA
print("a:\n", a)
print("Type: ", type(a))

a:
 Match ID
0              Dimitrov G.
1                Raonic M.
2            Kecmanovic M.
3               Millman J.
4              Uchiyama Y.
5                 Kudla D.
6                Chardy J.
7                Murray A.
8               Kyrgios N.
9              Tsonga J.W.
10            De Minaur A.
11             Thompson J.
12             Dimitrov G.
13             Uchiyama Y.
14            Nishikori K.
15               Raonic M.
16               Chardy J.
17             Medvedev D.
18            De Minaur A.
19             Tsonga J.W.
20               Chardy J.
21            Nishikori K.
22             Medvedev D.
23             Tsonga J.W.
24            Nishikori K.
25             Medvedev D.
26            Nishikori K.
27        Bautista Agut R.
28                Jarry N.
29             Verdasco F.
               ...        
2204             Bedene A.
2205           Tsonga J.W.
2206           Tsonga J.W.
2207          Mannarino A.
2208          Kukushkin M.
2209           

You can also select multiple columns. In this case you need to use a list.

Note that the result is a dataframe.

In [18]:
wl = df[ ["Winner","Loser"] ] #PRENDO INSIEMI DI COLONNE, OTTENGO UN DATAFRAME, NON UNA SERIE

wl.head()

Attributes         Winner        Loser
Match ID                              
0             Dimitrov G.  Nishioka Y.
1               Raonic M.    Bedene A.
2           Kecmanovic M.     Mayer L.
3              Millman J.  Sandgren T.
4             Uchiyama Y.   Humbert U.

#### Example (Addition)
We can add a new column by passing a NumPy array (or a list, Pandas Series, and more) or a single number.

In case of a list/array, the length of the array must equal the number of rows (otherwise a ValueError exception is raised)!

In [19]:
print("df.shape:", df.shape)

         # note that len gives us the number of rows
df['meaning-less']   = np.ones( len(df) ) #np.ones da vettori di uno (tra parentesi lunghezza che deve essere pari al dataframe)
df['meaning-less-2'] = 2
df['meaning-less-3'] = df['B365W']*2

df.head()

df.shape: (2234, 36)


Attributes  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                        
0             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Attributes Surface      Round  Best of         Winner  ... B365L   PSW   PSL  \
Match ID                                               ...                     
0             Hard  1st Round        3    Dimitrov G.  ...  3.00  1.36  3.37   
1             Hard  1st Round        3      Raonic M.  ...  4.50  1.23  4.68   
2             Hard  1st Round        3  Kecmanovic M.  ...  2.25  1.67  2.32   
3             Hard  1st Round        3     Millman J.  ...  2.75  1.41  3.13   
4             Hard  1st Round        3    Uchiyama Y.  ...  1.44  2.73  1.51   

Attributes  MaxW  MaxL  AvgW  AvgL  meaning-less  meaning-less-2  \
Match ID                                                           
0           1.42  3.60  1.35  3.18           1.0               2   
1           1.27  4.84  1.22  4.26           1.0               2   
2           1.71  2.40  1.63  2.28           1.0               2   
3           1.45  3.20  1.40  2.95           1.0               2   
4           3.26  1.53  2.69  1.47           1.0               2   

Attributes  meaning-less-3  
Match ID                    
0                     2.72  
1                     2.36  
2                     3.14  
3                     2.80  
4                     5.24  

[5 rows x 39 columns]

#### Example (Deletion)

Method `drop()` can be used to remove a column. We need to specify axis=1 (axis=0 are the rows).
The built-in `del` can also be used.

`inplace` argument (default is False)  is common in several Pandas' functions that modify the DataFrame. 
If True, it says that the function has to modify the DataFrame itself instead of returning a new one.

In [20]:
        # drop returns a new dataframe
df = df.drop('meaning-less', axis=1)   
        # unless inplace is set to True
df.drop('meaning-less-2', axis=1, inplace=True) #Modifico la struttura dati così senza restituire un altra
        # alternative
del df['meaning-less-3']

df.head()

Attributes  ATP  Location              Tournament       Date  Series    Court  \
Match ID                                                                        
0             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
1             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
2             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
3             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   
4             1  Brisbane  Brisbane International 2018-12-31  ATP250  Outdoor   

Attributes Surface      Round  Best of         Winner  ... Lsets    Comment  \
Match ID                                               ...                    
0             Hard  1st Round        3    Dimitrov G.  ...   0.0  Completed   
1             Hard  1st Round        3      Raonic M.  ...   0.0  Completed   
2             Hard  1st Round        3  Kecmanovic M.  ...   0.0  Completed   
3             Hard  1st Round        3     Millman J.  ...   1.0  Completed   
4             Hard  1st Round        3    Uchiyama Y.  ...   0.0  Completed   

Attributes  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Match ID                                                      
0            1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
1            1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
2            1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
3            1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
4            2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 36 columns]

## Index

Index does not need to be in integer, and it does not need to be unique.
We can choose one of the column to be the index with function `set_index()`. 

Note that the old index is lost!

In [21]:
df.set_index("Location", inplace=True) # Alternative to df = df.set_index("Location") VOGLIO CHE USI COME INDEX DELLA RIGA CIO CHE C'E' IN COLONNA LOCATION

# Note the new name of the index
df.head()

Attributes  ATP              Tournament       Date  Series    Court Surface  \
Location                                                                      
Brisbane      1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane      1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane      1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane      1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane      1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   

Attributes      Round  Best of         Winner        Loser  ...  Lsets  \
Location                                                    ...          
Brisbane    1st Round        3    Dimitrov G.  Nishioka Y.  ...    0.0   
Brisbane    1st Round        3      Raonic M.    Bedene A.  ...    0.0   
Brisbane    1st Round        3  Kecmanovic M.     Mayer L.  ...    0.0   
Brisbane    1st Round        3     Millman J.  Sandgren T.  ...    1.0   
Brisbane    1st Round        3    Uchiyama Y.   Humbert U.  ...    0.0   

Attributes    Comment  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Location                                                                 
Brisbane    Completed   1.36   3.00  1.36  3.37  1.42  3.60  1.35  3.18  
Brisbane    Completed   1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
Brisbane    Completed   1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
Brisbane    Completed   1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
Brisbane    Completed   2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  

[5 rows x 35 columns]

## Question: List the tournament names

In [22]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 

df['Tournament']

0       Brisbane International
1       Brisbane International
2       Brisbane International
3       Brisbane International
4       Brisbane International
5       Brisbane International
6       Brisbane International
7       Brisbane International
8       Brisbane International
9       Brisbane International
10      Brisbane International
11      Brisbane International
12      Brisbane International
13      Brisbane International
14      Brisbane International
15      Brisbane International
16      Brisbane International
17      Brisbane International
18      Brisbane International
19      Brisbane International
20      Brisbane International
21      Brisbane International
22      Brisbane International
23      Brisbane International
24      Brisbane International
25      Brisbane International
26      Brisbane International
27      Qatar Exxon Mobil Open
28      Qatar Exxon Mobil Open
29      Qatar Exxon Mobil Open
                 ...          
2204           Open de Moselle
2205    

In [25]:
set( df['Tournament'] )

{'ABN AMRO World Tennis Tournament',
 'ASB Classic',
 'Abierto Mexicano',
 'Abierto Mexicano Mifel',
 'Antalya Open',
 'Argentina Open',
 'Australian Open',
 'BB&T Atlanta Open',
 'BMW Open',
 'BNP Paribas Open',
 'Brasil Open',
 'Brisbane International',
 'Citi Open',
 'Cordoba Open',
 'Croatia Open',
 'Delray Beach Open',
 'Dubai Tennis Championships',
 'Eastbourne International',
 'French Open',
 'Generali Open',
 'Geneva Open',
 'German Tennis Championships',
 'Grand Prix Hassan II',
 'Hall of Fame Championships',
 'Halle Open',
 'Hungarian Open',
 "Internazionali BNL d'Italia",
 'Lyon Open',
 'Maharashtra Open',
 'Mercedes Cup',
 'Millennium Estoril Open',
 'Monte Carlo Masters',
 'Mutua Madrid Open',
 'New York Open',
 'Open 13',
 'Open Banco Sabadell ',
 'Open Sud de France',
 'Open de Moselle',
 'Qatar Exxon Mobil Open',
 "Queen's Club Championships",
 'Rio Open',
 'Rogers Masters',
 'Rosmalen Grass Court Championships',
 'SkiStar Swedish Open',
 'Sofia Open',
 'Sony Ericsson O

You can do this because a pandas Series is iterable, and a python set can be build from any iterable.

You can iterate in two ways.

In [26]:
for v in df['Tournament']:
    print (v)

Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Brisbane International
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon Mobil Open
Qatar Exxon

Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
Suisse Open Gstaad
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tennis Championships
German Tenn

In [27]:
# similar to enumerate
for index,v in df['Tournament'].iteritems():
    print (index, v)

0 Brisbane International
1 Brisbane International
2 Brisbane International
3 Brisbane International
4 Brisbane International
5 Brisbane International
6 Brisbane International
7 Brisbane International
8 Brisbane International
9 Brisbane International
10 Brisbane International
11 Brisbane International
12 Brisbane International
13 Brisbane International
14 Brisbane International
15 Brisbane International
16 Brisbane International
17 Brisbane International
18 Brisbane International
19 Brisbane International
20 Brisbane International
21 Brisbane International
22 Brisbane International
23 Brisbane International
24 Brisbane International
25 Brisbane International
26 Brisbane International
27 Qatar Exxon Mobil Open
28 Qatar Exxon Mobil Open
29 Qatar Exxon Mobil Open
30 Qatar Exxon Mobil Open
31 Qatar Exxon Mobil Open
32 Qatar Exxon Mobil Open
33 Qatar Exxon Mobil Open
34 Qatar Exxon Mobil Open
35 Qatar Exxon Mobil Open
36 Qatar Exxon Mobil Open
37 Qatar Exxon Mobil Open
38 Qatar Exxon Mobil O

998 Millennium Estoril Open
999 Millennium Estoril Open
1000 Millennium Estoril Open
1001 Millennium Estoril Open
1002 Millennium Estoril Open
1003 Millennium Estoril Open
1004 Millennium Estoril Open
1005 Millennium Estoril Open
1006 Millennium Estoril Open
1007 Millennium Estoril Open
1008 Millennium Estoril Open
1009 Millennium Estoril Open
1010 Millennium Estoril Open
1011 Millennium Estoril Open
1012 Millennium Estoril Open
1013 Millennium Estoril Open
1014 BMW Open
1015 BMW Open
1016 BMW Open
1017 BMW Open
1018 BMW Open
1019 BMW Open
1020 BMW Open
1021 BMW Open
1022 BMW Open
1023 BMW Open
1024 BMW Open
1025 BMW Open
1026 BMW Open
1027 BMW Open
1028 BMW Open
1029 BMW Open
1030 BMW Open
1031 BMW Open
1032 BMW Open
1033 BMW Open
1034 BMW Open
1035 BMW Open
1036 BMW Open
1037 BMW Open
1038 BMW Open
1039 BMW Open
1040 BMW Open
1041 Mutua Madrid Open
1042 Mutua Madrid Open
1043 Mutua Madrid Open
1044 Mutua Madrid Open
1045 Mutua Madrid Open
1046 Mutua Madrid Open
1047 Mutua Madrid Open

1748 Suisse Open Gstaad
1749 Suisse Open Gstaad
1750 Suisse Open Gstaad
1751 Suisse Open Gstaad
1752 Suisse Open Gstaad
1753 Suisse Open Gstaad
1754 Suisse Open Gstaad
1755 Suisse Open Gstaad
1756 Suisse Open Gstaad
1757 Suisse Open Gstaad
1758 Suisse Open Gstaad
1759 Suisse Open Gstaad
1760 Suisse Open Gstaad
1761 Suisse Open Gstaad
1762 Suisse Open Gstaad
1763 Suisse Open Gstaad
1764 German Tennis Championships
1765 German Tennis Championships
1766 German Tennis Championships
1767 German Tennis Championships
1768 German Tennis Championships
1769 German Tennis Championships
1770 German Tennis Championships
1771 German Tennis Championships
1772 German Tennis Championships
1773 German Tennis Championships
1774 German Tennis Championships
1775 German Tennis Championships
1776 German Tennis Championships
1777 German Tennis Championships
1778 German Tennis Championships
1779 German Tennis Championships
1780 German Tennis Championships
1781 German Tennis Championships
1782 German Tennis Cha

Pandas allows to do better then using python sets.

In [28]:
df['Tournament'].unique() #NON PRENDE VALORI REPLICATI
# see https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.unique.html

array(['Brisbane International', 'Qatar Exxon Mobil Open',
       'Maharashtra Open', 'ASB Classic', 'Sydney International',
       'Australian Open', 'Cordoba Open', 'Open Sud de France',
       'Sofia Open', 'Argentina Open', 'New York Open',
       'ABN AMRO World Tennis Tournament', 'Delray Beach Open', 'Open 13',
       'Rio Open', 'Abierto Mexicano', 'Dubai Tennis Championships',
       'Brasil Open', 'BNP Paribas Open', 'Sony Ericsson Open',
       "U.S. Men's Clay Court Championships", 'Grand Prix Hassan II',
       'Monte Carlo Masters', 'Open Banco Sabadell ', 'Hungarian Open',
       'Millennium Estoril Open', 'BMW Open', 'Mutua Madrid Open',
       "Internazionali BNL d'Italia", 'Geneva Open', 'Lyon Open',
       'French Open', 'Rosmalen Grass Court Championships',
       'Mercedes Cup', 'Halle Open', "Queen's Club Championships",
       'Antalya Open', 'Eastbourne International', 'Wimbledon',
       'SkiStar Swedish Open', 'Hall of Fame Championships',
       'Croatia Open

In [29]:
df['Tournament'].value_counts() #FREQUENZA DI UN CERTO CAMPO
# see: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html

Australian Open                                 127
Wimbledon                                       127
US Open                                         127
French Open                                     127
BNP Paribas Open                                 95
Sony Ericsson Open                               95
Rogers Masters                                   55
Western & Southern Financial Group Masters       55
Monte Carlo Masters                              55
Internazionali BNL d'Italia                      55
Mutua Madrid Open                                55
Winston-Salem Open at Wake Forest University     47
Citi Open                                        47
Open Banco Sabadell                              47
Queen's Club Championships                       31
Halle Open                                       31
Dubai Tennis Championships                       31
ABN AMRO World Tennis Tournament                 31
Grand Prix Hassan II                             31
Abierto Mexi

In [30]:
# Note it returns a Series
for tournament, count in df['Tournament'].value_counts().iteritems():
    print ("During the",tournament, "there were", count, "matches.")

During the Australian Open there were 127 matches.
During the Wimbledon there were 127 matches.
During the US Open there were 127 matches.
During the French Open there were 127 matches.
During the BNP Paribas Open there were 95 matches.
During the Sony Ericsson Open there were 95 matches.
During the Rogers Masters there were 55 matches.
During the Western & Southern Financial Group Masters there were 55 matches.
During the Monte Carlo Masters there were 55 matches.
During the Internazionali BNL d'Italia there were 55 matches.
During the Mutua Madrid Open there were 55 matches.
During the Winston-Salem Open at Wake Forest University there were 47 matches.
During the Citi Open there were 47 matches.
During the Open Banco Sabadell  there were 47 matches.
During the Queen's Club Championships there were 31 matches.
During the Halle Open there were 31 matches.
During the Dubai Tennis Championships there were 31 matches.
During the ABN AMRO World Tennis Tournament there were 31 matches.
Duri

# Question: Find player with most wins

In [31]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 

winners = df['Winner'].value_counts()
most_winner = winners.index[0]
most_winner

'Medvedev D.'

## A focus  on Series

- **Series** is a one-dimensional array-like object containing a sequence of values (of similar types to NumPy types) and an associated array of data labels, called its *index*.

It provides functionalities similar to that of python lists and python dictionaries.

As for Dataframes, Series have an index. The default index is made by numbers from 0-1, otherwise we can specify it, and we can use arbitrary labels as index.


In [32]:
import pandas as pd

pds = pd.Series([4, 7, -5, -3])

print(pds)

0    4
1    7
2   -5
3   -3
dtype: int64


In [33]:
pds = pd.Series( [4, 7, -5, -3], 
                 index=["George", "John", "Paul", "Ringo"])

# Series my have a name and its index may also have a name
pds.name = "The Beatles"
pds.index.name = "Member"

print(pds)

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64


### Indexing and slicing

Series indexing works analogously to Python list if you want to access by position, or similar to python dictionaries if you want to access by index value.

In [34]:
# access by position
print (pds[2])
print ()

# access by index value
print (pds['Paul'])

-5

-5


In [35]:
# Fancy indexing by position
print (pds[ [0,3] ])
print ()

# Fancy indexing by index calue
p = pds[  ['John', 'Ringo']  ]
print ( p )
print ()

# note: they return pands series
print (type(p))

Member
George    4
Ringo    -3
Name: The Beatles, dtype: int64

Member
John     7
Ringo   -3
Name: The Beatles, dtype: int64

<class 'pandas.core.series.Series'>


In [36]:
# We also have Boolean Indexing
print (pds[ [True, False, False, True] ])

Member
George    4
Ringo    -3
Name: The Beatles, dtype: int64


## Slicing, loc and iloc

Pandas tries to understand whether you are using a position or an index value. (What if they are both integers?)

To avoid confusion, one good recommendation is to use `loc` and `iloc`.

#### Example
As in Python, we can use slicing with positions. **Right extreme is NOT included**.

Recommendation: use `iloc`

In [37]:
#Iloc -> Integer location
#loc -> nomi righe, nomi colonne

print ( pds )
print ()

print ( pds[1:3] )
print ()

print ( pds.iloc[1:3] )
print ()

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64

Member
John    7
Paul   -5
Name: The Beatles, dtype: int64

Member
John    7
Paul   -5
Name: The Beatles, dtype: int64



#### Example
But we can also use slicing with labels. **Right extreme is included**. Index must be sorted !

Recommendation: use `loc`

In [38]:
#SI PUO' FARE SOLO SU INDICI ORDINATI

print ( pds['George':'P'] )     # 'P' is not present
print ()

print ( pds['George':'Paul'] ) #ESTREMO DESTRO E' INCLUSO
print ()

print ( pds.loc['George':'Paul'] )


Member
George    4
John      7
Name: The Beatles, dtype: int64

Member
George    4
John      7
Paul     -5
Name: The Beatles, dtype: int64

Member
George    4
John      7
Paul     -5
Name: The Beatles, dtype: int64


As for a dictionary you can check for presence.

In [40]:
print("'John' in pds:", "John" in pds )
print("'Mark' in pds:", "Mark" in pds )

# try
# pds["Mark"] # KeyError exception if not present

'John' in pds: True
'Mark' in pds: False


### Series are mutable

We can change its values with an assignment (also with slicing). 

In [41]:
print(pds)
print()

pds['Paul'] = 42

pds['George':'Paul'] = 5

print(pds)

Member
George    4
John      7
Paul     -5
Ringo    -3
Name: The Beatles, dtype: int64

Member
George    5
John      5
Paul      5
Ringo    -3
Name: The Beatles, dtype: int64


In [42]:
pds['Gennaro'] = 12 # It's a honor be part of this group

print(pds)

Member
George      5
John        5
Paul        5
Ringo      -3
Gennaro    12
Name: The Beatles, dtype: int64


### Filtering + Boolean Indexing

We can filter entries of a Series.

(You have the same in NumPy's Boolean indexing).

#### Example
Here we get only rows with a positive value. 

In [43]:
a = pds > 0
print("pds > 0\n", a)
print()

print("pds[ pds > 0]:\n", 
       pds[ pds > 0 ]  ) # get only positive values of pds 
print("Fab Four are back!")

pds > 0
 Member
George      True
John        True
Paul        True
Ringo      False
Gennaro     True
Name: The Beatles, dtype: bool

pds[ pds > 0]:
 Member
George      5
John        5
Paul        5
Gennaro    12
Name: The Beatles, dtype: int64
Fab Four are back!


In [44]:
# Bye-bye gennaro

psd = pds.drop('Gennaro')

### Operations
We can perform NumPy operations on a Series. 

In [45]:
import numpy as np

print("pds*2:\n", pds*2 )
print()

print("np.exp( pds ):\n", np.exp( pds ) )

pds*2:
 Member
George     10
John       10
Paul       10
Ringo      -6
Gennaro    24
Name: The Beatles, dtype: int64

np.exp( pds ):
 Member
George        148.413159
John          148.413159
Paul          148.413159
Ringo           0.049787
Gennaro    162754.791419
Name: The Beatles, dtype: float64


### Series from a dictionary

Series are so close to a dictionary that you can create one from a dictionary.

In [46]:
sdata = {'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
pds = pd.Series(sdata)

print(pds)

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64


When passing also an index, this is used to filter matching entries of the dictionary.

In [47]:
states = ['California', 'Ohio', 'Oregon', 'Texas']
pds = pd.Series(sdata, index=states) # California is not in sdata

print(pds) # Welcome missing values

California        NaN
Ohio          35000.0
Oregon        16000.0
Texas         71000.0
dtype: float64


Note that `'Utah'` was not included, and that a special value `NaN` is used for the index `California`.

Missing data is common, (e.g., movies withouth ratings), and usually `NaN` is used to represent them. 

It is possible to use `isnull` to find null values, and to replace them.

In [48]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

print (pd.isnull(obj))
print ()

obj [ pd.isnull(obj) ] = 0.0

print (obj)

Ohio          False
Texas         False
California     True
dtype: bool

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


Alternatively one can use the `fillna` method.

In [49]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

obj2 = obj.fillna(0.0) # returns a new data frame
print (obj)
print ()
print (obj2)

Ohio          35000.0
Texas         71000.0
California        NaN
dtype: float64

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


In [50]:
obj = pd.Series({'Ohio': 35000, 'Texas': 71000, 
                 'Oregon': 16000, 'Utah': 5000},
               index=['Ohio', 'Texas', 'California'])

obj.fillna(0.0, inplace=True) # inplace modification
print (obj)

Ohio          35000.0
Texas         71000.0
California        0.0
dtype: float64


### Allignment by index

A useful Series feature for many applications is that it automatically aligns by index
label in arithmetic operations. 

In [51]:
pds1 = pd.Series({'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000})  # California is missing
  
pds2 = pd.Series({'California': 40000, 'Texas': 555, 'Oregon': 111, 'Utah': 222}) # Ohio is missing

pds1 + pds2

California        NaN
Ohio              NaN
Oregon        16111.0
Texas         71555.0
Utah           5222.0
dtype: float64

In [52]:
pds1 = pd.Series({'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000})  # California is missing
  
pds2 = pd.Series({'California': 40000, 'Texas': 555, 'Oregon': 111, 'Utah': 222}) # Ohio is missing

pds1.add(pds2, fill_value=0)

California    40000.0
Ohio          35000.0
Oregon        16111.0
Texas         71555.0
Utah           5222.0
dtype: float64

## Question: List the player names, and count the number of matches they had


In [53]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 

players = df['Winner'].value_counts() + df['Loser'].value_counts()
players.sort_values()

King K.                 2.0
Brands D.               2.0
Celikbilek A.           2.0
Serdarusic N.           2.0
Statham J.              2.0
Griekspoor T.           2.0
Ofner S.                2.0
Brooksby J.             2.0
Krawietz K.             2.0
Diez S.                 2.0
Lee D.H.                2.0
Lamasine T.             2.0
Moroni G.               2.0
Domingues J.            3.0
Donaldson J.            3.0
Otte O.                 3.0
Gombos N.               3.0
Polmans M.              3.0
Kuhn N.                 3.0
Mmoh M.                 3.0
Giraldo S.              3.0
Cachin P.               3.0
Giannessi A.            3.0
Brown D.                3.0
Benchetrit E.           3.0
Menendez-Maceiras A.    4.0
Pospisil V.             4.0
Clarke J.               4.0
Bolelli S.              4.0
Uchiyama Y.             4.0
                       ... 
Polansky P.             NaN
Robredo T.              NaN
Rodionov J.             NaN
Rola B.                 NaN
Rosol L.            

In [54]:
players = df['Winner'].value_counts().add(df['Loser'].value_counts(), fill_value=0)

# Note: you can sort !
players.sort_values(ascending=False)

Medvedev D.           70.0
Tsitsipas S.          57.0
Nadal R.              56.0
Paire B.              55.0
Pella G.              53.0
Schwartzman D.        53.0
Federer R.            52.0
Bautista Agut R.      52.0
Struff J.L.           51.0
Berrettini M.         51.0
Goffin D.             51.0
Djokovic N.           50.0
Fritz T.              49.0
Zverev A.             49.0
Auger-Aliassime F.    49.0
Albot R.              47.0
Ramos-Vinolas A.      46.0
Sousa J.              46.0
Simon G.              46.0
Thiem D.              45.0
Verdasco F.           45.0
Chardy J.             45.0
Basilashvili N.       45.0
Wawrinka S.           44.0
Fucsovics M.          44.0
Herbert P.H.          44.0
Thompson J.           44.0
Monfils G.            44.0
Nishikori K.          43.0
Hurkacz H.            43.0
                      ... 
King D.                1.0
Kirkin E.              1.0
Skugor F.              1.0
Krstin P.              1.0
Krueger M.             1.0
Kubler J.              1.0
K

## Row Selection, Addition, and Deletion

Very similar to column operation , with a different way of specifying rows.

More details at https://pandas.pydata.org/pandas-docs/stable/indexing.html.

#### Example (Selection)
We can access rows by using the index value in the special `loc` attribute.

Note that the result is a dataframe.

In [23]:
df.set_index("Location", inplace=True) # Alternative to df = df.set_index("Location")

df.loc['Brisbane'] # get all the mathing rows

ATP              Tournament       Date  Series    Court Surface  \
Location                                                                    
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-02  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-03  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-04  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-04  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-05  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-05  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-06  ATP250  Outdoor    Hard   

                  Round  Best of         Winner          Loser  ...  Lsets  \
Location                                                        ...          
Brisbane      1st Round        3    Dimitrov G.    Nishioka Y.  ...    0.0   
Brisbane      1st Round        3      Raonic M.      Bedene A.  ...    0.0   
Brisbane      1st Round        3  Kecmanovic M.       Mayer L.  ...    0.0   
Brisbane      1st Round        3     Millman J.    Sandgren T.  ...    1.0   
Brisbane      1st Round        3    Uchiyama Y.     Humbert U.  ...    0.0   
Brisbane      1st Round        3       Kudla D.       Fritz T.  ...    1.0   
Brisbane      1st Round        3      Chardy J.    Struff J.L.  ...    1.0   
Brisbane      1st Round        3      Murray A.   Duckworth J.  ...    0.0   
Brisbane      1st Round        3     Kyrgios N.    Harrison R.  ...    1.0   
Brisbane      1st Round        3    Tsonga J.W.  Kokkinakis T.  ...    0.0   
Brisbane      1st Round        3   De Minaur A.     Popyrin A.  ...    0.0   
Brisbane      1st Round        3    Thompson J.        Bolt A.  ...    0.0   
Brisbane      2nd Round        3    Dimitrov G.     Millman J.  ...    0.0   
Brisbane      2nd Round        3    Uchiyama Y.      Edmund K.  ...    0.0   
Brisbane      2nd Round        3   Nishikori K.       Kudla D.  ...    0.0   
Brisbane      2nd Round        3      Raonic M.  Kecmanovic M.  ...    0.0   
Brisbane      2nd Round        3      Chardy J.     Kyrgios N.  ...    1.0   
Brisbane      2nd Round        3    Medvedev D.      Murray A.  ...    0.0   
Brisbane      2nd Round        3   De Minaur A.    Thompson J.  ...    0.0   
Brisbane      2nd Round        3    Tsonga J.W.      Daniel T.  ...    0.0   
Brisbane  Quarterfinals        3      Chardy J.    Uchiyama 

#### Example (Selection)
We can access rows by using a list of index values.

In [24]:
df.loc[ ['Paris','London'] ] 

ATP   Tournament       Date      Series    Court Surface  \
Location                                                             
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-26  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
Paris      32  French Open 2019-05-27  Grand Slam  Outdoor    Clay   
...       ...          ...        ...         ...      ...     ...   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-05  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-06  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-08  Grand Slam  Outdoor   Grass   
London     39    Wimbledon 2019-07-10  Grand Slam  Outdoor   Grass   
London     39    Wim

#### Example (Selection)
Rows can be selected by using integer location with the special `iloc` attribute.

In [25]:
df.iloc[2]

ATP                                1
Tournament    Brisbane International
Date             2018-12-31 00:00:00
Series                        ATP250
Court                        Outdoor
Surface                         Hard
Round                      1st Round
Best of                            3
Winner                 Kecmanovic M.
Loser                       Mayer L.
WRank                            131
LRank                             56
WPts                             433
LPts                             895
W1                                 6
L1                                 3
W2                                 6
L2                                 1
W3                               NaN
L3                               NaN
W4                               NaN
L4                               NaN
W5                               NaN
L5                               NaN
Wsets                              2
Lsets                              0
Comment                    Completed
B

#### Example (Selection)
Slicing over the rows with `iloc`.

In [26]:
df.iloc[1:10]

ATP              Tournament       Date  Series    Court Surface  \
Location                                                                    
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2018-12-31  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   
Brisbane    1  Brisbane International 2019-01-01  ATP250  Outdoor    Hard   

              Round  Best of         Winner          Loser  ...  Lsets  \
Location                                                    ...          
Brisbane  1st Round        3      Raonic M.      Bedene A.  ...    0.0   
Brisbane  1st Round        3  Kecmanovic M.       Mayer L.  ...    0.0   
Brisbane  1st Round        3     Millman J.    Sandgren T.  ...    1.0   
Brisbane  1st Round        3    Uchiyama Y.     Humbert U.  ...    0.0   
Brisbane  1st Round        3       Kudla D.       Fritz T.  ...    1.0   
Brisbane  1st Round        3      Chardy J.    Struff J.L.  ...    1.0   
Brisbane  1st Round        3      Murray A.   Duckworth J.  ...    0.0   
Brisbane  1st Round        3     Kyrgios N.    Harrison R.  ...    1.0   
Brisbane  1st Round        3    Tsonga J.W.  Kokkinakis T.  ...    0.0   

            Comment  B365W  B365L   PSW   PSL  MaxW  MaxL  AvgW  AvgL  
Location                                                               
Brisbane  Completed   1.18   4.50  1.23  4.68  1.27  4.84  1.22  4.26  
Brisbane  Completed   1.57   2.25  1.67  2.32  1.71  2.40  1.63  2.28  
Brisbane  Completed   1.40   2.75  1.41  3.13  1.45  3.20  1.40  2.95  
Brisbane  Completed   2.62   1.44  2.73  1.51  3.26  1.53  2.69  1.47  
Brisbane  Completed   2.62   1.44  2.80  1.49  2.85  1.55  2.70  1.47  
Brisbane  Completed   2.10   1.66  2.23  1.72  2.26  1.74  2.19  1.68  
Brisbane  Completed   1.28   3.50  1.38  3.29  1.39  3.60  1.34  3.26  
Brisbane  Completed   1.40   2.75  1.47  2.87  1.50  3.16  1.44  2.80  
Brisbane  Completed   2.25   1.57  2.27  1.70  2.42  1.71  2.27  1.64  

[9 rows x 35 columns]

#### Example (Addition)
We can add a new row by using `loc` attribute.

Pandas also `append` has function (see https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.append.html).

In [27]:
# Make a smaller copy
small_df = pd.DataFrame( df[ ['Surface', 'Winner', 'Loser'] ] ) 
small_df = small_df.loc[['London', 'Paris']]

small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Nadal R.']
small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Federer R.'] # This is a replace

small_df

Surface              Winner             Loser
Location                                              
London     Grass       Tipsarevic J.       Nishioka Y.
London     Grass         Anderson K.      Herbert P.H.
London     Grass            Lopez F.          Giron M.
London     Grass    Bautista Agut R.      Gojowczyk P.
London     Grass           Opelka R.        Stebe C.M.
London     Grass         Wawrinka S.      Bemelmans R.
London     Grass            Mayer L.         Gulbis E.
London     Grass        Khachanov K.         Kwon S.W.
London     Grass           Darcis S.         Zverev M.
London     Grass            Kudla D.         Jaziri M.
London     Grass            Seppi A.          Jarry N.
London     Grass  Auger-Aliassime F.       Pospisil V.
London     Grass         Djokovic N.  Kohlschreiber P.
London     Grass            Haase R.        Kovalik J.
London     Grass         Fabbiano T.      Tsitsipas S.
London     Grass            Pella G.          Copil M.
London     Grass           Cuevas P.        Dzumhur D.
London     Grass         Karlovic I.      Arnaboldi A.
London     Grass          Popyrin A.  Carreno Busta P.
London     Grass           Vesely J.         Zverev A.
London     Grass          Hurkacz H.        Lajovic D.
London     Grass         Verdasco F.      Majchrzak K.
London     Grass         Medvedev D.        Lorenzi P.
London     Grass            Paire B.      Londero J.I.
London     Grass          Humbert U.        Monfils G.
London     Grass           Chardy J.         Klizan M.
London     Grass           Raonic M.    Gunneswaran P.
London     Grass           Moutet C.       Dimitrov G.
London     Grass           Edmund K.          Munar J.
London     Grass       Granollers M.         Sonego L.
...          ...                 ...               ...
Paris       Clay          Federer R.           Ruud C.
Paris       Clay            Mayer L.          Mahut N.
Paris       Clay        Londero J.I.         Moutet C.
Paris       Clay            Nadal R.         Goffin D.
Paris       Clay          Fognini F.  Bautista Agut R.
Paris       Clay           Zverev A.        Lajovic D.
Paris       Clay        Tsitsipas S.     Krajinovic F.
Paris       Clay         Struff J.L.          Coric B.
Paris       Clay         Wawrinka S.       Dimitrov G.
Paris       Clay         Djokovic N.         Caruso S.
Paris       Clay            Thiem D.         Cuevas P.
Paris       Clay          Monfils G.          Hoang A.
Paris       Clay      Del Potro J.M.       Thompson J.
Paris       Clay        Khachanov K.         Klizan M.
Paris       Clay          Federer R.          Mayer L.
Paris       Clay         Wawrinka S.      Tsitsipas S.
Paris       Clay            Nadal R.      Londero J.I.
Paris       Clay        Nishikori K.          Paire B.
Paris       Clay         Djokovic N.       Struff J.L.
Paris       Clay            Thiem D.        Monfils G.
Paris       Clay           Zverev A.        Fognini F.
Paris       Clay        Khachanov K.    Del Potro J.M.
Paris       Clay          Federer R.       Wawrinka S.
Paris       Clay            Nadal R.      Nishikori K.
Paris       Clay         Djokovic N.         Zverev A.
Paris       Clay            Thiem D.      Khachanov K.
Paris       Clay            Nadal R.        Federer R.
Paris       Clay            Thiem D.       Djokovic N.
Paris       Clay            Nadal R.          Thiem D.
Mestre      Clay          Claudio L.        Federer R.

[255 rows x 3 columns]

#### Example (Deletion)
We can remove a row (or more) with `drop()` function.

In [28]:
small_df.drop("Mestre", inplace=True) # axis=0 is the default
small_df

Surface              Winner             Loser
Location                                              
London     Grass       Tipsarevic J.       Nishioka Y.
London     Grass         Anderson K.      Herbert P.H.
London     Grass            Lopez F.          Giron M.
London     Grass    Bautista Agut R.      Gojowczyk P.
London     Grass           Opelka R.        Stebe C.M.
London     Grass         Wawrinka S.      Bemelmans R.
London     Grass            Mayer L.         Gulbis E.
London     Grass        Khachanov K.         Kwon S.W.
London     Grass           Darcis S.         Zverev M.
London     Grass            Kudla D.         Jaziri M.
London     Grass            Seppi A.          Jarry N.
London     Grass  Auger-Aliassime F.       Pospisil V.
London     Grass         Djokovic N.  Kohlschreiber P.
London     Grass            Haase R.        Kovalik J.
London     Grass         Fabbiano T.      Tsitsipas S.
London     Grass            Pella G.          Copil M.
London     Grass           Cuevas P.        Dzumhur D.
London     Grass         Karlovic I.      Arnaboldi A.
London     Grass          Popyrin A.  Carreno Busta P.
London     Grass           Vesely J.         Zverev A.
London     Grass          Hurkacz H.        Lajovic D.
London     Grass         Verdasco F.      Majchrzak K.
London     Grass         Medvedev D.        Lorenzi P.
London     Grass            Paire B.      Londero J.I.
London     Grass          Humbert U.        Monfils G.
London     Grass           Chardy J.         Klizan M.
London     Grass           Raonic M.    Gunneswaran P.
London     Grass           Moutet C.       Dimitrov G.
London     Grass           Edmund K.          Munar J.
London     Grass       Granollers M.         Sonego L.
...          ...                 ...               ...
Paris       Clay            Paire B.  Carreno Busta P.
Paris       Clay          Federer R.           Ruud C.
Paris       Clay            Mayer L.          Mahut N.
Paris       Clay        Londero J.I.         Moutet C.
Paris       Clay            Nadal R.         Goffin D.
Paris       Clay          Fognini F.  Bautista Agut R.
Paris       Clay           Zverev A.        Lajovic D.
Paris       Clay        Tsitsipas S.     Krajinovic F.
Paris       Clay         Struff J.L.          Coric B.
Paris       Clay         Wawrinka S.       Dimitrov G.
Paris       Clay         Djokovic N.         Caruso S.
Paris       Clay            Thiem D.         Cuevas P.
Paris       Clay          Monfils G.          Hoang A.
Paris       Clay      Del Potro J.M.       Thompson J.
Paris       Clay        Khachanov K.         Klizan M.
Paris       Clay          Federer R.          Mayer L.
Paris       Clay         Wawrinka S.      Tsitsipas S.
Paris       Clay            Nadal R.      Londero J.I.
Paris       Clay        Nishikori K.          Paire B.
Paris       Clay         Djokovic N.       Struff J.L.
Paris       Clay            Thiem D.        Monfils G.
Paris       Clay           Zverev A.        Fognini F.
Paris       Clay        Khachanov K.    Del Potro J.M.
Paris       Clay          Federer R.       Wawrinka S.
Paris       Clay            Nadal R.      Nishikori K.
Paris       Clay         Djokovic N.         Zverev A.
Paris       Clay            Thiem D.      Khachanov K.
Paris       Clay            Nadal R.        Federer R.
Paris       Clay            Thiem D.       Djokovic N.
Paris       Clay            Nadal R.          Thiem D.

[254 rows x 3 columns]

In [29]:
# Make a smaller copy
small_df = pd.DataFrame( df[ ['Surface', 'Winner', 'Loser'] ] ) 
small_df = small_df.loc[['London', 'Paris']]

small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Nadal R.']
small_df.loc['Mestre'] = ['Clay', 'Claudio L.', 'Federer R.'] # This is a replace

# Remove all matching rows
small_df.drop(["London", "Paris"], inplace=True)
small_df

Surface      Winner       Loser
Location                                
Mestre      Clay  Claudio L.  Federer R.

### Creating a DataFrame

There are many ways to construct a DataFrame, though one of the most common is from a dictionary of equal-length lists (or NumPy arrays).

In [30]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],
        'population': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}

df = pd.DataFrame(data) 

df.index.name = "Progressive"   # Set index name
df.columns.name = 'Attributes'  # Set columns name

df # Default index is 0...N-1 

Attributes    state  year  population
Progressive                          
0              Ohio  2000         1.5
1              Ohio  2001         1.7
2              Ohio  2002         3.6
3            Nevada  2001         2.4
4            Nevada  2002         2.9
5            Nevada  2003         3.2

In [31]:
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],
        'population': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}

df = pd.DataFrame(data,
                 index = ['xxx1','xxx2','xxx3','xxx4','xxx5','xxx6']) 

df.index.name   = "Custom ID"   # Set index name
df.columns.name = 'Attributes'  # Set columns name

df

Attributes   state  year  population
Custom ID                           
xxx1          Ohio  2000         1.5
xxx2          Ohio  2001         1.7
xxx3          Ohio  2002         3.6
xxx4        Nevada  2001         2.4
xxx5        Nevada  2002         2.9
xxx6        Nevada  2003         3.2

## From pandas to numpy

If you prefer working with numpy

In [32]:
m = df.values
print (type(m))
print ()

print (m)

<class 'numpy.ndarray'>

[['Ohio' 2000 1.5]
 ['Ohio' 2001 1.7]
 ['Ohio' 2002 3.6]
 ['Nevada' 2001 2.4]
 ['Nevada' 2002 2.9]
 ['Nevada' 2003 3.2]]


# Question: Find Most unexpected result by Nadal

In [33]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [34]:
nadal_winner = df[ df['Winner']=='Nadal R.' ]
nadal_winner = nadal_winner[['Winner', 'B365W']] #B365W valore sulla vittoria
nadal_winner.columns = ['Player','bet'] # see also rename function
nadal_winner.head()

Player   bet
149  Nadal R.  1.05
218  Nadal R.  1.06
241  Nadal R.  1.10
252  Nadal R.  1.14
260  Nadal R.  1.04

In [35]:
nadal_loser = df[ df['Loser']=='Nadal R.' ]
nadal_loser = nadal_loser[['Loser', 'B365L']] #B356L valore sulla sconfitta
nadal_loser.columns = ['Player','bet'] 
nadal_loser.head()

Player   bet
265  Nadal R.  2.15
542  Nadal R.  1.22
703  Nadal R.  2.75
911  Nadal R.  1.08
958  Nadal R.  1.36

In [36]:
nadal_matches = pd.concat([nadal_winner, nadal_loser])
nadal_matches

Player    bet
149   Nadal R.  1.050
218   Nadal R.  1.060
241   Nadal R.  1.100
252   Nadal R.  1.140
260   Nadal R.  1.040
263   Nadal R.  1.160
532   Nadal R.  1.050
672   Nadal R.  1.040
684   Nadal R.  1.080
690   Nadal R.  1.070
701   Nadal R.  1.160
895   Nadal R.  1.080
902   Nadal R.  1.030
908   Nadal R.  1.020
940   Nadal R.  1.030
949   Nadal R.  1.140
955   Nadal R.  1.060
1075  Nadal R.  1.140
1088  Nadal R.  1.050
1092  Nadal R.  1.200
1129  Nadal R.  1.050
1141  Nadal R.  1.050
1145  Nadal R.  1.050
1148  Nadal R.  1.250
1150  Nadal R.  1.440
1232  Nadal R.  1.005
1271  Nadal R.  1.002
1306  Nadal R.  1.030
1319  Nadal R.  1.005
1326  Nadal R.  1.020
1329  Nadal R.  1.120
1331  Nadal R.  1.200
1555  Nadal R.  1.020
1595  Nadal R.  1.250
1611  Nadal R.  1.160
1615  Nadal R.  1.020
1625  Nadal R.  1.120
1925  Nadal R.  1.080
1941  Nadal R.  1.080
1946  Nadal R.  1.120
1949  Nadal R.  1.050
1950  Nadal R.  1.400
2115  Nadal R.  1.020
2147  Nadal R.  1.040
2158  Nadal R.  1.020
2172  Nadal R.  1.070
2176  Nadal R.  1.050
2178  Nadal R.  1.060
2179  Nadal R.  1.160
265   Nadal R.  2.150
542   Nadal R.  1.220
703   Nadal R.  2.750
911   Nadal R.  1.080
958   Nadal R.  1.360
1094  Nadal R.  1.100
1627  Nadal R.  1.720

In [37]:
nadal_matches.sort_values(by='bet',ascending=False, inplace=True)
nadal_matches.head()

Player   bet
703   Nadal R.  2.75
265   Nadal R.  2.15
1627  Nadal R.  1.72
1150  Nadal R.  1.44
1950  Nadal R.  1.40

In [38]:
match_id = nadal_matches.index[0]
print (df.loc[match_id])

ATP                            19
Location             Indian Wells
Tournament       BNP Paribas Open
Date          2019-03-16 00:00:00
Series               Masters 1000
Court                     Outdoor
Surface                      Hard
Round                  Semifinals
Best of                         3
Winner                 Federer R.
Loser                    Nadal R.
WRank                           4
LRank                           2
WPts                         4600
LPts                         8365
W1                            NaN
L1                            NaN
W2                            NaN
L2                            NaN
W3                            NaN
L3                            NaN
W4                            NaN
L4                            NaN
W5                            NaN
L5                            NaN
Wsets                         NaN
Lsets                         NaN
Comment                  Walkover
B365W                        1.44
B365L         

# Question: how many times the player with the best ranking won the match?


In [39]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [40]:
df.columns

Index(['ATP', 'Location', 'Tournament', 'Date', 'Series', 'Court', 'Surface',
       'Round', 'Best of', 'Winner', 'Loser', 'WRank', 'LRank', 'WPts', 'LPts',
       'W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5', 'Wsets',
       'Lsets', 'Comment', 'B365W', 'B365L', 'PSW', 'PSL', 'MaxW', 'MaxL',
       'AvgW', 'AvgL'],
      dtype='object')

In [41]:
w_gt_r = df['WRank']>df['LRank']

w_gt_r.head()

0    False
1    False
2     True
3    False
4     True
dtype: bool

In [42]:
w_gt_r = w_gt_r.astype(int)

w_gt_r.head()

0    0
1    0
2    1
3    0
4    1
dtype: int32

In [43]:
total_wins = w_gt_r.sum()

total_wins

849

In [44]:
total_matches, _ = df.shape

total_matches

2234

In [45]:
print ("The success rate of the best ranked player is", round(100.0*total_wins/total_matches,2), "%" )

The success rate of the best ranked player is 38.0 %


# Question: how much would you gain or lose by always betting 10€ on the best ranked player?

In [ ]:
import pandas as pd

dataset_file = './datasets/tennis/2019.xlsx'

df = pd.read_excel(dataset_file) 

In [ ]:
w_gt_r = df['WRank']>df['LRank']

In [ ]:
# decompose the following to understand each singles setps
gains = ( (df['B365W'][w_gt_r]-1.0) * 5.0).sum()

In [ ]:
losses = (~w_gt_r).sum() * 5.0

In [ ]:
total = gains - losses

print ("If always betting on the best ranked, the profit would be", total)

## Additional useful manipulation

Suppose, for some reason, you want to invert name/surname order.

That is, suppose you want to apply the same &custom* function to every element of a dataframe/series.

In [ ]:
def my_fun (x):
    tokens = x.split()
    tokens = tokens[::-1]
    new_x  = ' '.join(tokens)
    return new_x

# note the re-assiggnment
df['Winner'] = df['Winner'].map(my_fun)

df['Winner']

## Question: What is the surface with longest matches on average (more games)?

In [ ]:
df.columns

In [ ]:
sub_df = df[ ['W1', 'L1', 'W2', 'L2', 'W3', 'L3', 'W4', 'L4', 'W5', 'L5'] ]
sub_df.head()

In [ ]:
sub_df = sub_df.fillna(0.0)
sub_df.head()

In [ ]:
sub_df.sum()

In [ ]:
sub_df.sum(axis=1)

In [ ]:
df['games'] = sub_df.sum(axis=1)

In [ ]:
df[ ['Surface','games'] ].groupby('Surface').mean()

## Pivot Tables

Similar to groupby, also columsn are grouped.

 - See https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pivot_table.html
 
The function `pivot_table` has the following main parameters.

| Param name | Description|
|-|:-|
| values  | The column to be aggregated |
| index   | Column names used to create rows of the Pivot Table |
| columns | Column names used to create cols of the Pivot Table  |
| aggfunc | Aggregation function (e.g., `np.sum`)


# Question: Find the most successful player by surface 

In [ ]:
import pandas as pd
dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 


df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len)

In [ ]:
df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )

In [ ]:
wins = df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
wins.head()

In [ ]:
wins.columns

In [ ]:
for surface in wins.columns:
    print (wins[surface].sort_values(ascending=False).head(1))
    print ()

# Question: Find the player with the best success rate  by surface

In [1]:
import pandas as pd
dataset_file = './datasets/tennis/2019.xlsx'
df = pd.read_excel(dataset_file) 

In [2]:
wins = df.pivot_table(values="ATP", # irrelevant
               index="Winner",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
wins.head()

Surface             Clay  Grass  Hard
Winner                               
Albot R.               7      1    19
Anderson K.            0      3     8
Andreozzi G.           2      1     3
Andujar P.             9      0     5
Auger-Aliassime F.    12      9    10

In [3]:
losses = df.pivot_table(values="ATP", # irrelevant
               index="Loser",
               columns="Surface",
               aggfunc=len,
               fill_value=0.0 )
losses.head()

Surface       Clay  Grass  Hard
Loser                          
Albot R.         6      3    11
Altmaier D.      1      0     0
Anderson K.      0      2     2
Andreev A.       0      0     1
Andreozzi G.     8      2     4

In [4]:
total = wins + losses
total.head()

Surface       Clay  Grass  Hard
Albot R.      13.0    4.0  30.0
Altmaier D.    NaN    NaN   NaN
Anderson K.    0.0    5.0  10.0
Andreev A.     NaN    NaN   NaN
Andreozzi G.  10.0    3.0   7.0

In [5]:
total = wins.add(losses, fill_value=0)
total.head()

Surface       Clay  Grass  Hard
Albot R.      13.0    4.0  30.0
Altmaier D.    1.0    0.0   0.0
Anderson K.    0.0    5.0  10.0
Andreev A.     0.0    0.0   1.0
Andreozzi G.  10.0    3.0   7.0

In [8]:
wins.loc['Nadal R.']

Surface
Clay     21
Grass     5
Hard     23
Name: Nadal R., dtype: int64

In [7]:
losses.loc['Nadal R.']

Surface
Clay     3
Grass    1
Hard     3
Name: Nadal R., dtype: int64

In [ ]:
total.loc['Nadal R.']

In [ ]:
success_rate = wins / total

success_rate.head()

In [ ]:
success_rate = wins / total
success_rate.fillna(0.0, inplace=True)
success_rate.head()

In [ ]:
success_rate.loc['Nadal R.']

In [ ]:
success_rate.loc['Thiem D.']

In [ ]:
for surface in success_rate:  # note: you do not need to use .columns
    print (success_rate[surface].sort_values(ascending=False).head(1))
    print ()

## Other Useful stuff



### DataFrame Joins

#### Many-to-One join
Consider the following two DataFrames.

In [ ]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a', 'c', 'a', 'a', 'b'], 
                    'data1': range(7)})

df2 = pd.DataFrame({'key': ['a', 'b', 'd'], 
                    'data2': range(3)})

print(df1)
print(df2)

The data in df1 has multiple rows labeled a and b, whereas df2 has only one row for each value in the key column. 

The operation 

```
pd.merge(df1, df2, on='key')
```

will produce a merged DataFrame with key the column to join on. 


In [ ]:
pd.merge(df1, df2, on='key')

Notice that 
- data2 of df2 is replicated everywhere key 'a' and 'b' occur in df1
- rows with key 'c' in df1 and 'd' in df2 are not present in df

The latter behavior is called **inner** join: the keys in the result are the intersection, or the common set found in both tables.

Other possible approaches are: 
- **left**: keys of left DataFrame are kept
- **right**: keys of right DataFrame are kept
- **outer**: keys of both DataFrames are kept

and can be chosen by setting parameter *how*

#### left

In [ ]:
pd.merge(df1, df2, on='key', how='left')

#### right

In [ ]:
pd.merge(df1, df2, on='key', how='right')

#### outer

In [ ]:
pd.merge(df1, df2, on='key', how='outer')

#### Many-to-Many join
Many-to-many merges happens when the same key has more than one occurrence in both DataFrames.

Many-to-many joins form the Cartesian product of the rows having the same key. 

See the following example:

In [ ]:
df1 = pd.DataFrame({'key': ['b', 'b', 'a',  'a', 'b'], 
                    'data1': range(5)})
                    
df2 = pd.DataFrame({'key': ['a', 'b', 'a', 'b', 'd'],
                    'data2': range(5)})
                    
pd.merge(df1, df2, on='key', how='left')

Note that, since there were three 'b' rows in df1 and two in df2, there are six 'b' rows in the result.

The merge can be performed also respect to more than one variable. To determine which key combinations will appear in the result depending on the choice of merge method, **think of the multiple keys as forming an array of tuples** to be used as a single join key.

# References

 - **Python for Data Analysis**. O'Reilly. Wes McKinney.
   - Section 5.1 (Indexing, Selection, and Filtering)
   - Section 5.2 (Arithmetic and Data Alignment Function Application and Mapping Sorting and Ranking)
   - Section 5.3 (Unique Values, Value Counts, and Membership)
   - Section 10.4 (Pivot Tables)
   - Section 14.5
   - https://github.com/wesm/pydata-book
